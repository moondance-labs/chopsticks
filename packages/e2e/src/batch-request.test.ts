import { WebSocket } from 'ws'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'

import networks from './networks.js'

describe('Batch request', async () => {
  const { chain, url, teardown } = await networks.acala()
  const wsClient = new WebSocket(url)

  beforeAll(async () => {
    await new Promise<void>((resolve) => {
      wsClient.on('open', () => resolve())
    })
  })

  afterAll(async () => {
    wsClient.close()
    await teardown()
  })

  it('batch getStorage', async () => {
    const blockNumber = chain.head.number

    const isFinish = new Promise<void>((resolve) => {
      wsClient.once('finished', () => {
        resolve()
      })
    })

    wsClient.on('message', (data: any) => {
      const response = JSON.parse(data)

      expect(response).toMatchSnapshot()
      wsClient.emit('finished')
    })

    wsClient.send(
      JSON.stringify([
        {
          method: 'chain_getBlockHash',
          params: [blockNumber],
          id: 2,
          jsonrpc: '2.0',
        },
        {
          method: 'not_found',
          params: [''],
          id: 3,
          jsonrpc: '2.0',
        },
      ]),
    )

    await isFinish
  })
})
