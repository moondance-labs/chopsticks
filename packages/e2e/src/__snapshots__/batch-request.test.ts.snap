// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Batch request > batch getStorage 1`] = `
[
  {
    "id": 2,
    "jsonrpc": "2.0",
    "result": "0xb5297d01adb0964d5195f9f17a3cf6e99ef8622e71863456eeb9296d5681292b",
  },
  {
    "error": {
      "code": -32601,
      "message": "Method not found: not_found",
    },
    "id": 3,
    "jsonrpc": "2.0",
  },
]
`;
